plugins/OnlineSales.Plugin.SendGrid/Controllers/SendgridController.cs

```diff
@@ -0,0 +1,281 @@
+using OnlineSales.Data;
+using OnlineSales.Entities;
+using OnlineSales.Interfaces;
+using OnlineSales.Plugin.SendGrid.Data;
+using OnlineSales.Plugin.SendGrid.DTOs;
+using OnlineSales.Plugin.SendGrid.Entities;
+using OnlineSales.Plugin.SendGrid.Exceptions;
+using SendGrid.Helpers.EventWebhook;
+using Serilog;
+
+namespace OnlineSales.Plugin.SendGrid.Controllers;
+
+[Route("api/[controller]")]
+public class SendgridController : ControllerBase
+{
+    private static readonly int BatchSize = 10;
+
+    private static readonly DateTime UnixTimestampZeroDate = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+
+    private readonly SendgridDbContext dbContext;
+
+    private readonly IContactService contactService;
+
+    public SendgridController(SendgridDbContext dbContext, EsDbContext esDbContext, IConfiguration configuration, IContactService contactService)
+    {
+        this.dbContext = dbContext;
+        this.contactService = contactService;
+    }
+
+    [HttpPost]
+    [Authorize]
+    [Route("import")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status500InternalServerError)]
+    public async Task<ActionResult> Import([FromBody] List<MessageEventDto> records)
+    {
+        return await AddEvents(records);
+    }
+
+    [HttpPost]
+    [AllowAnonymous]
+    [Route("webhook")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status500InternalServerError)]
+    public async Task<ActionResult> AddMessageEvents()
+    {
+        var signature = Request.Headers[RequestValidator.SIGNATURE_HEADER].First();
+        var timestamp = Request.Headers[RequestValidator.TIMESTAMP_HEADER].First();
+        var body = await GetRequestBody(HttpContext.Request.Body);
+
+        if (signature == null || timestamp == null || !Verify(body, signature, timestamp))
+        {
+            throw new SendGridApiException("Webhook event can't be verified");
+        }
+
+        var records = JsonConvert.DeserializeObject<List<WebhookEventDto>>(body);
+
+        if (records == null)
+        {
+            throw new SendGridApiException("Webhook event body can't be deserialized");
+        }
+
+        return await AddEvents(records);
+    }
+
+    private async Task<Dictionary<Contact, List<T>>> CreateNonExistedContacts<T>(IEnumerable<IGrouping<string, T>> emailAndRecords)
+    {
+        var contactRecords = new Dictionary<Contact, List<T>>();
+        var nonExistedContacts = new List<Contact>();
+
+        int position = 0;
+
+        while (position < emailAndRecords.Count())
+        {
+            var batch = emailAndRecords.Skip(position).Take(BatchSize);
+            var existedContacts = dbContext.Contacts!.Where(c => batch.Select(b => b.Key).Contains(c.Email)).ToList();
```

Comment:
@IgorKostyukov it is better to make a LINQ query which returns Dictionary<string, Contact> where key = email instead of a simple list (consider using ToDictionary instead of ToList). In such a way you will not need to run full collection scan as you do it not on line 92. Instead, you will be able to use dictionary.TryGet method to check if contact exists or not.